File: src/C#/BehavioralPatterns/State/VendingMachine/IdleState.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace VendingMachine
{
    public class IdleState : State
    {
        public IdleState(VendingMachine vendingMachine) : base(vendingMachine)
        {
            Console.WriteLine("IDLE - Wait for product selection");
        }

        public override void InsertMoney(decimal amount)
        {
            Console.WriteLine("Please select a product before inserting any money.");
        }

        public override void SelectProduct(string productCode)
        {
            var selectedProduct = vendingMachine.Products.FirstOrDefault(p => p.Code == productCode);

            if (selectedProduct.Stock == 0)
            {
                Console.WriteLine($"The product code: {selectedProduct.Code} is out of stock");
                return;
            }

            vendingMachine.SelectedProductCode = selectedProduct.Code;
            Console.WriteLine($"Product: {selectedProduct.Code} with price: {selectedProduct.Price} selected.");
            vendingMachine.SetState(new PaymentState(vendingMachine));
        }

        public override void DispenseProduct() => Console.WriteLine("Select a product first.");

        public override void Cancel() => Console.WriteLine("There is no selected product or payment in process to cancel.");

        public override void Refill(List<Product> products)
        {
            vendingMachine.Products = products;
            Console.WriteLine($"Total amount of products: {vendingMachine.Products.Sum(p => p.Stock)}");
        }
    }
}

File: src/JavaScript/BehavioralPatterns/State/VendingMachine/idleState.js

const State = require("./state");
const PaymentState = require("./paymentState");

class IdleState extends State {
    constructor(vendingMachine) {
        super(vendingMachine);
        console.log("IDLE - Wait for product selection");
    }

    insertMoney(amount) {
        console.log("Please select a product before inserting any money.");
    }

    selectProduct(productCode) {
        const selectedProduct = this.vendingMachine.products.find(p => p.code === productCode);

        if (!selectedProduct || selectedProduct.stock === 0) {
            console.log(`The product code: ${productCode} is out of stock`);
            return;
        }

        this.vendingMachine.selectedProductCode = productCode;
        console.log(`Product: ${productCode} with price: ${selectedProduct.price} selected.`);
        this.vendingMachine.setState(new PaymentState(this.vendingMachine));
    }

    dispenseProduct() {
        console.log("Select a product first.");
    }

    cancel() {
        console.log("There is no selected product or payment in process to cancel.");
    }

    refill(products) {
        this.vendingMachine.products = products;
        console.log(`Total amount of products: ${this.vendingMachine.products.reduce((total, p) => total + p.stock, 0)}`);
    }
}

module.exports = IdleState;